### PR TITLE
feat: transparent proxy

### PR DESCRIPTION
**Pull request by 3062**
### Issues associated with this PR

Issues #1928

### Solutions
You should show your solutions about the issues in your PR, including the overall solutions, details and the changes. At this time, Chinese is allowed to describe these.

### UT result
Unit Test is needed if the code is changed, your unit test should cover boundary cases, corner cases, and some exceptional cases. And you need to show the UT result.

### Benchmark
If your code involves the processing of every request, you should give the Benchmark Result.

### Code Style
+ Make sure `Goimports` has run
+ Show `Golint` result

### CODE REVIEW

File: pkg/filter/listener/originaldst/originaldst.go [hunk 1]
```diff
@@ -32,32 +32,46 @@ const (
...
 	port := p0*256 + p1
 
-	ip := addr.Multiaddr[4:8]
+	ips := addr.Multiaddr[4:8]
+	ip := fmt.Sprintf("%d.%d.%d.%d", ips[0], ips[1], ips[2], ips[3])
 
 	return ip, port, nil
 }
+
+func getTransparentProxyAddr(conn net.Conn) (string, int, error) {
```
Comment (taoyuanyuan) [inline pkg/filter/listener/originaldst/originaldst.go:66]: Simplify all TransparentProxy to TProxy.

File: pkg/network/connection.go [hunk 2]
```diff
@@ -1133,7 +1134,16 @@ func (cc *clientConnection) connect() (event api.ConnectionEvent, err error) {
...
 		return api.ConnectFailed, errors.New("ClientConnection RemoteAddr is nil")
 	}
-	cc.rawConnection, err = net.DialTimeout(cc.network, cc.RemoteAddr().String(), timeout)
+
+	dialer := &net.Dialer{
+		Timeout:   timeout,
+	}
+
+	if cc.mark != 0 {
+		dialer.Control = SockMarkControl
```
Comment (taoyuanyuan) [inline pkg/network/connection.go:1143]: maybe we can use closure to simplify logic.

File: pkg/filter/listener/originaldst/factory.go [hunk 3]
```diff
@@ -19,28 +19,49 @@ package originaldst
...
+	Type            string `json:"type"`
+}
+
+func CreateOriginalDstConfig(conf map[string]interface{}) (OriginalDstConfig, error) {
+	b, _ := json.Marshal(conf)
+	cfg := OriginalDstConfig{}
+	if err := json.Unmarshal(b, &cfg); err != nil {
+		return cfg, err
+	}
+
```
Comment (taoyuanyuan) [inline pkg/filter/listener/originaldst/factory.go:54]: judge whether the parameter is legal.

File: pkg/filter/listener/originaldst/originaldst.go [hunk 4]
```diff
@@ -32,32 +32,50 @@ const (
 	IP6T_SO_ORIGINAL_DST = 80
 )
 
-func getOriginalAddr(conn net.Conn) ([]byte, int, error) {
-	tc := conn.(*net.TCPConn)
+func getRedirectAddr(conn net.Conn) (string, int, error) {
+	tc, ok := conn.(*net.TCPConn)
+
```
Comment (taoyuanyuan) [inline pkg/filter/listener/originaldst/originaldst.go:37]: delete blank line.

File: pkg/filter/listener/originaldst/factory.go [hunk 5]
```diff
@@ -42,29 +37,13 @@ func init() {
...
-
-	if cfg.Type == "" {
-		cfg.Type = Redirect
-	} else if cfg.Type != Redirect && cfg.Type != TProxy {
-		return cfg, errors.New("listener filter type unrecognized")
-	}
-
-	return cfg, nil
+	FallbackToLocal bool               `json:"fallback_to_local"`
+	Type            v2.OriginalDstType `json:"type"`
```
Comment (taoyuanyuan) [inline pkg/filter/listener/originaldst/factory.go:41]: 不需要这个配置了呢，配置在listener里面的

File: pkg/types/network.go [hunk 6]
```diff
@@ -110,10 +110,10 @@ type Listener interface {
 	SetPerConnBufferLimitBytes(limitBytes uint32)
 
 	// Set if listener should use original dst
-	SetUseOriginalDst(use bool)
+	SetUseOriginalDst(use v2.OriginalDstType)
```
Comment (taoyuanyuan) [inline pkg/types/network.go:113]: SetUseOriginalDst rename to SetOriginalDst and  add a new interface named GetOriginalDst

File: pkg/types/network.go [hunk 7]
```diff
@@ -110,10 +110,10 @@ type Listener interface {
...
 
 	// Get if listener should use original dst
-	UseOriginalDst() bool
+	GetUseOriginalDst() bool
```
Comment (taoyuanyuan) [inline pkg/types/network.go:116]: GetUseOriginalDst rename to UseOriginalDst

File: pkg/types/network.go (hunk 6 above)
Comment (taoyuanyuan) [reply pkg/types/network.go:113]: use rename to type

File: pkg/network/connection.go (hunk 2 above)
Comment (3062) [reply pkg/network/connection.go]: passing error through closure

**Review by wangfakang — APPROVED** [review]

**Review by taoyuanyuan — APPROVED** [review]
LGTM

**Pull request merged**